File: Data Structures & Algorithms/partition-labels/submission-0.cpp
class Solution {
public:
    vector<int> partitionLabels(string s) {
    map<char, int> mp;
    vector <int> ans;
    ans.push_back(-1);
    for(int i = 0;i<s.size();i++){
        mp[s[i]] = i;
    }

    int l = 0;
    int r = mp[s[0]];
    if (r == s.size()-1){
        ans.push_back(r);
        return ans;
    }
    while (l<s.size()){
        if(s[l] == s[r]){
            if (l!=r)
                l++;
            else{
                ans.push_back(r);
                l++;
                r = max(mp[s[l]],r);
            }
        }
        else{
            if(mp[s[l]]<r)
                l++;
            r=max(mp[s[l]],r);
        }
    }
    vector <int> vec;
    for(int i = 1;i<ans.size();i++){
        vec.push_back(ans[i]-ans[i-1]);
    }
    return vec;
    }
};

File: Data Structures & Algorithms/capacity-to-ship-packages-within-d-days/submission-0.py
class Solution:
    def shipWithinDays(self, w: List[int], days: int) -> int:
        def check(n):
            t = 1
            s = 0
            for i in w:
                if s + i > n:
                    t+=1
                    s = 0

                s+=i
            print(days,t,n)
            if (t > days):
                return False
            return True

        lo = max(w)
        hi = sum(w)
        ans = 0
        while (lo<=hi):
            m = (lo+hi)//2
            if check(m):
                ans = m
                hi = m-1

            else:
                lo = m + 1

        return ans



